aws/resource_aws_eks_node_group.go

```diff
@@ -374,19 +412,39 @@ func resourceAwsEksNodeGroupUpdate(d *schema.ResourceData, meta interface{}) err
 		}
 	}
 
-	if d.HasChanges("release_version", "version") {
+	if d.HasChanges("launch_template", "release_version", "version") {
 		input := &eks.UpdateNodegroupVersionInput{
 			ClientRequestToken: aws.String(resource.UniqueId()),
 			ClusterName:        aws.String(clusterName),
 			Force:              aws.Bool(d.Get("force_update_version").(bool)),
 			NodegroupName:      aws.String(nodeGroupName),
 		}
 
+		if v := d.Get("launch_template").([]interface{}); len(v) > 0 {
+			input.LaunchTemplate = expandEksLaunchTemplateSpecification(v)
+
+			// When returning Launch Template information, the API returns all
+			// fields. Since both the id and name are saved to the Terraform
+			// state for drift detection and the API returns the following
+			// error if both are present during update:
+			// InvalidParameterException: Either provide launch template ID or launch template name in the request.
+
+			// Remove the name if there are no changes, to prefer the ID.
+			if input.LaunchTemplate.Id != nil && input.LaunchTemplate.Name != nil && !d.HasChange("launch_template.0.name") {
+				input.LaunchTemplate.Name = nil
+			}
+
+			// Otherwise, remove the ID, but only if both are present still.
+			if input.LaunchTemplate.Id != nil && input.LaunchTemplate.Name != nil && !d.HasChange("launch_template.0.id") {
+				input.LaunchTemplate.Id = nil
+			}
+		}
+
 		if v, ok := d.GetOk("release_version"); ok && d.HasChange("release_version") {
 			input.ReleaseVersion = aws.String(v.(string))
 		}
 
-		if v, ok := d.GetOk("version"); ok {
+		if v, ok := d.GetOk("version"); ok && d.HasChange("version") {
```

Comment:
interesting, my only follow-up to this would be if we need to check for `!d.HasChange("launch_template")` here along with the check for changes in `version` to satisfy the condition that the "API does not allow specifying the Kubernetes version when Launch Templates are being updated" since we're still inside the outer `if d.HasChanges("launch_template", "release_version", "version")` at this point? an example i'm thinking is in the case where a node group `version` is updated/rolled-back to a previous one. though I'm not certain if that's a valid condition..

Reply:
ah i see `TestAccAWSEksNodeGroup_LaunchTemplate_Version` would cover something similar..at first i didn't catch it's the ssm_parameter data source that changes things downstream 👍